src/base/path.cpp

```diff
@@ -0,0 +1,310 @@
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "path.h"
+
+#include <QDataStream>
+#include <QDir>
+#include <QFileInfo>
+#include <QMimeDatabase>
+#include <QRegularExpression>
+
+Path::Path(const QString &pathStr)
+    : m_pathStr {QDir::cleanPath(pathStr)}
+{
+}
+
+Path::Path(const std::string &pathStr)
+    : Path(QString::fromStdString(pathStr))
+{
+}
+
+Path::Path(const char pathStr[])
+    : Path(QString::fromLatin1(pathStr))
+{
+}
+
+bool Path::isValid() const
+{
+    if (isEmpty())
+        return false;
+
+#if defined(Q_OS_WIN)
+    const QRegularExpression regex {QLatin1String("[:?\"*<>|]")};
+#elif defined(Q_OS_MACOS)
+    const QRegularExpression regex {QLatin1String("[\\0:]")};
+#else
+    const QRegularExpression regex {QLatin1String("[\\0]")};
+#endif
+    return !m_pathStr.contains(regex);
+}
+
+bool Path::isEmpty() const
+{
+    return m_pathStr.isEmpty();
+}
+
+bool Path::isAbsolute() const
+{
+    return QDir::isAbsolutePath(m_pathStr);
+}
+
+bool Path::isRelative() const
+{
+    return QDir::isRelativePath(m_pathStr);
+}
+
+bool Path::exists() const
+{
+    return !isEmpty() && QFileInfo::exists(m_pathStr);
+}
+
+Path Path::rootItem() const
+{
+    const int slashIndex = m_pathStr.indexOf(QLatin1Char('/'));
+    if (slashIndex < 0)
+        return *this;
+
+    if (slashIndex == 0) // *nix absolute path
+        return createUnchecked(QLatin1String("/"));
+
+    return createUnchecked(m_pathStr.left(slashIndex));
+}
+
+Path Path::parentPath() const
+{
+    const int slashIndex = m_pathStr.lastIndexOf(QLatin1Char('/'));
+    if (slashIndex == -1)
+        return {};
+
+    if (slashIndex == 0) // *nix absolute path
+        return (m_pathStr.size() == 1) ? Path() : createUnchecked(QLatin1String("/"));
+
+    return createUnchecked(m_pathStr.left(slashIndex));
+}
+
+Path Path::filename() const
+{
+    const int slashIndex = m_pathStr.lastIndexOf('/');
+    if (slashIndex == -1)
+        return *this;
+
+    return createUnchecked(m_pathStr.mid(slashIndex + 1));
+}
+
+QString Path::extension() const
+{
+    const QString suffix = QMimeDatabase().suffixForFileName(m_pathStr);
+    return (suffix.isEmpty() ? suffix : (QLatin1String(".") + suffix));
+}
+
+bool Path::hasExtension(const QString &ext, Qt::CaseSensitivity caseSensitivity) const
+{
+    return (extension().compare(ext, caseSensitivity) == 0);
+}
+
+bool Path::isEqualTo(const Path &other, const Qt::CaseSensitivity caseSensitivity) const
```

Comment:
> Is it a good idea to expose this signature instead?

It seems to me that the concept of greaterThan/lessThan is not very applicable to the path. Only the concept of equal (not equal) makes sense.
Anyway, as I said above, `It does not pretend to be general purpose solution`. If such a method is really required in the future we'll add it.

Reply:
As a user, it is a bit confusing to see both `isEqualTo()` and `operator==()` exists side by side. Maybe just make this function always compare case insensitively?

Reply:
> Maybe just make this function always compare case insensitively?

Then it will be wrong on case sensitive systems. Wouldn't it be better to make it system dependent?
Really we need both comparison way:
1. Accurate comparison (always case sensitive)
2. System dependent comparison (case insensitive on appropriate systems)

We should use system dependent comparison everywhere we need to know whether two paths point to the same file/folder.
Accurate comparison is still useful when we compare old and new paths before moving/renaming since it is valid to change only case of filename even on case insensitive systems.
So question is which of them should be an `operator==()`, and which an additional function?

Reply:
> So question is which of them should be an `operator==()`, and which an additional function?

I think `operator==()` should compare in platform dependent way. If we need accurate comparison we can compare them as strings.